File: examples/serving/inference_csharp/DataNode.cs
using System;
using System.Collections.Generic;
using Microsoft.ML.OnnxRuntime.Tensors;

namespace inference_csharp
{
  
    /// <summary>
    /// data for one user's history
    /// </summary>
    public class History_Node
    {
        public long user_id { get; set; }
        public long[] item_seq { get; set; } = null!;
        public long[] time_seq { get; set; } = null!;
        public bool has_time { get; set; }
    }

    /// <summary>
    /// data for one item's features
    /// </summary>
    public class Feature_Node
    {
        public long item_id { get; set; }
        public long[] item_feats { get; set; } = null!;
        public Feature_Node()
        {
            item_id = 0;
            item_feats = new long[GlobalVar.n_features];
            for(int i = 0; i < GlobalVar.n_features; i++)
            {
                item_feats[i] = 0;
            }
        }
    }

    /// <summary>
    /// one test case
    /// </summary>
    public class Test_Node
    {
        public long user_id { get; set; }
        public long item_id { get; set; }
    }

    /// <summary>
    /// collect for all user history and test case
    /// </summary>
    public class Data
    {
        public History_Node[] user_history { get; set; } = null!;
        public Feature_Node[] item_features { get; set; } = null!;
        public List<Test_Node> test_set { get; set; } = null!;
        public Data()
        {
            user_history = new History_Node[GlobalVar.n_users];
            test_set = new List<Test_Node>();
        }
    }

    /// <summary>
    /// input for one batch, default value is all zero
    /// </summary>
    public class Batch_Input
    {
        // 'user_id' 'item_id' 'item_features' 'item_seq' 'item_seq_features'
        public Tensor<long> user_id { get; set; } = null!;
        public Tensor<long> item_id { get; set; } = null!;
        public Tensor<long> item_features { get; set; } = null!;
        public Tensor<long> item_seq { get; set; } = null!;
        public Tensor<long> item_seq_len { get; set; } = null!;
        public Tensor<long> item_seq_features { get; set; } = null!;
        public Tensor<long> time_seq { get; set; } = null!;
        public Batch_Input(int batch_size)
        {
            user_id = new DenseTensor<long>(new[] { batch_size });
            item_id = new DenseTensor<long>(new[] { batch_size });
            item_seq = new DenseTensor<long>(new[] { batch_size, GlobalVar.max_seq_len });
            item_seq_len = new DenseTensor<long>(new[] { batch_size });
            time_seq = new DenseTensor<long>(new[] { batch_size, GlobalVar.max_seq_len });
            for (int i = 0; i < batch_size; i++)
            {
                for(int j = 0; j < GlobalVar.max_seq_len; j++)
                {
                    item_seq[i, j] = 0;
                }
            }
            for (int i = 0; i < batch_size; i++)
            {
                item_seq_len[i] = 0;
            }
            for (int i = 0; i < batch_size; i++)
            {
                for(int j = 0; j < GlobalVar.max_seq_len; j++)
                {
                    time_seq[i, j] = 0;
                }
            }
        }
    }
}

File: examples/serving/inference_java/src/main/java/com/example/DataNode.java
package com.example;
import java.util.List;  
  
public class DataNode {  
    public static class HistoryNode {  
        public long userId;  
        public long[] itemSeq;  
        public long[] timeSeq;  
        public boolean hasTime;  
    }  
  
    public static class FeatureNode {  
        public long itemId;  
        public long[] itemFeats;  
  
        public FeatureNode() {  
            itemId = 0;  
            itemFeats = new long[Main.nFeatures];  
            for (int i = 0; i < Main.nFeatures; i++) {  
                itemFeats[i] = 0;  
            }  
        }  
    }  
  
    public static class TestNode {  
        public long userId;  
        public long itemId;  
    }  
  
    public static class Data {  
        public HistoryNode[] userHistory;  
        public FeatureNode[] itemFeatures;  
        public List<TestNode> testSet;  
  
        public Data() {  
            userHistory = new HistoryNode[Main.nUsers];  
        }  
    }  
  
    public static class BatchInput {  
        public long[] userId;  
        public long[] itemId;  
        public long[][] itemFeatures;  
        public long[][] itemSeq;  
        public long[] itemSeqLen;  
        public long[][][] itemSeqFeatures;  
        public long[][] timeSeq;  
  
        public BatchInput(int batchSize) {  
            userId = new long[batchSize];  
            itemId = new long[batchSize];  
            itemSeq = new long[batchSize][Main.maxSeqLen];  
            itemSeqLen = new long[batchSize];  
            timeSeq = new long[batchSize][Main.maxSeqLen];
            for (int i = 0; i < batchSize; i++) {  
                for (int j = 0; j < Main.maxSeqLen; j++) {  
                    itemSeq[i][j] = 0;  
                    timeSeq[i][j] = 0;  
                }
                itemSeqLen[i] = 0;
            }  
        }  
    }  
}